### ISSUE TITLE
qsort_r not present in alpine linux

### ISSUE THREAD

**Issue by pgmillon** [libgit2/libgit2#3438]
When trying to build libgit2 on alpine linux :

```
CMakeFiles/libgit2_clar.dir/src/util.c.o: In function `git__qsort_r':
util.c:(.text+0x137f): undefined reference to `qsort_r'
collect2: error: ld returned 1 exit status
CMakeFiles/libgit2_clar.dir/build.make:11753: recipe for target 'libgit2_clar' failed
make[2]: *** [libgit2_clar] Error 1
CMakeFiles/Makefile2:104: recipe for target 'CMakeFiles/libgit2_clar.dir/all' failed
make[1]: *** [CMakeFiles/libgit2_clar.dir/all] Error 2
Makefile:126: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by ethomson** [comment]
Ugh.  That suggests to me that they've gone and ripped things out of glibc...?

Have they at least set some `#define ALPINE_LINUX` that can be checked for?

**Comment by carlosmn** [comment]
Let's not check for a particular distribution. We have checks for which extra libraries we need for certain things, like the timers. The approach to take here would be to check for `qsort_r()` in libc and set the define if not (since we need to deal with other OSs which do this as well).

**Comment by arthurschreiber** [comment]
> Ugh. That suggests to me that they've gone and ripped things out of glibc...?

Quick googling shows they don't use glibc but musl libc instead. And that does not seem to have `qsort_r`.

**Comment by runelabs** [comment]
I have the same problem with building libgit2 under musl-libc. Musl is used on Alpine Linux, OpenEmbedded. hardened Gentoo Linux variants etc.

Musl uses _smoothsort_ as its `qsort` implementation, but does not provide `qsort_r`.

References:
* https://github.com/bminor/musl
* http://calmerthanyouare.org/2013/05/31/qsort-shootout.html
* http://www.keithschwarz.com/smoothsort/
* http://www.etalabs.net/compare_libcs.html

**Comment by ethomson** [comment]
Thanks - this should be fixed via #3446

**Comment by Kochise** [comment]
Fyi, qsort has been added to Alpine on 2021/09/22

https://gitlab.alpinelinux.org/alpine/aports/-/commit/bf5bbfdbf780092f387b7abe401fbfceda90c84d

Profit.
